src/telink/hal/zigbee_zcl.c

```diff
@@ -0,0 +1,221 @@
+  u16 *in_cluster_ptr = in_clusters;
+  u16 *out_cluster_ptr = out_clusters;
+  zcl_specClusterInfo_t *cluster_info_ptr = cluster_infos;
+  zclAttrInfo_t *attr_table_ptr = attr_tables;
+
+  for (hal_zigbee_endpoint *endpoint = endpoints;
+       endpoint < endpoints + hal_endpoints_cnt; endpoint++) {
+    // Initialize endpoint descriptors
+    endpoint_desc_ptr->endpoint = endpoint->endpoint;
+    endpoint_desc_ptr->app_profile_id = endpoint->profile_id;
+    endpoint_desc_ptr->app_dev_id = endpoint->device_id;
+    endpoint_desc_ptr->app_dev_ver = endpoint->device_version;
+    endpoint_desc_ptr->app_in_cluster_lst = in_cluster_ptr;
+    endpoint_desc_ptr->app_out_cluster_lst = out_cluster_ptr;
+
+    zcl_specClusterInfo_t *endpoint_first_cluster_ptr = cluster_info_ptr;
+    for (hal_zigbee_cluster *cluster = endpoint->clusters;
+         cluster < endpoint->clusters + endpoint->cluster_count; cluster++) {
+      if (cluster->is_server) {
+        *in_cluster_ptr = cluster->cluster_id;
+        in_cluster_ptr++;
+        endpoint_desc_ptr->app_in_cluster_count++;
+      } else {
+        *out_cluster_ptr = cluster->cluster_id;
+        out_cluster_ptr++;
+        endpoint_desc_ptr->app_out_cluster_count++;
+      }
+      if (cluster->cluster_id == ZCL_CLUSTER_OTA) {
+        // OTA cluster is handled separately
+        continue;
+      }
+      // Initialize cluster info
+      cluster_info_ptr->clusterId = cluster->cluster_id;
+      cluster_info_ptr->manuCode = 0;
+      cluster_info_ptr->attrTbl = attr_table_ptr;
+      cluster_info_ptr->attrNum = 0;
+      cluster_info_ptr->clusterRegisterFunc =
+          get_register_func_by_cluster_id(cluster->cluster_id);
+      cluster_info_ptr->clusterAppCb =
+          get_cmd_callback_by_cluster_id(cluster->cluster_id);
+      for (hal_zigbee_attribute *attr = cluster->attributes;
+           attr < cluster->attributes + cluster->attribute_count; attr++) {
+        // Copy attribute to global table
+        attr_table_ptr->id = attr->attribute_id;
+        attr_table_ptr->type = attr->data_type_id;
+        attr_table_ptr->access =
+            ACCESS_CONTROL_READ | ACCESS_CONTROL_REPORTABLE;
+        if (attr->flag == ATTR_WRITABLE) {
+          attr_table_ptr->access |= ACCESS_CONTROL_WRITE;
+        }
+        attr_table_ptr->data = attr->value;
+        attr_table_ptr++;
+        cluster_info_ptr->attrNum++;
+      }
+
+      cluster_info_ptr++;
+    }
+    af_endpointRegister(endpoint->endpoint, endpoint_desc_ptr, zcl_rx_handler,
+                        NULL);
+    zcl_register(endpoint->endpoint,
+                 cluster_info_ptr - endpoint_first_cluster_ptr,
+                 endpoint_first_cluster_ptr);
+    endpoint_desc_ptr++;
+  }
+}
+
+void hal_zigbee_notify_attribute_changed(uint8_t endpoint, uint8_t cluster_id,
+                                         uint16_t attribute_id) {
+  report_handler(); // Trigger reporting if needed
+}
```

Comment:
There's an issue caused by this change. For more details and a fix, see https://github.com/romasku/tuya-zigbee-switch/pull/185

Reply:
Yes, it does.